dealii-policy-compatibility.md

```diff
@@ -0,0 +1,42 @@
+and should be considered when filling out this form.
+
+Please, provide information on your compability status for each mandatory policy, and if possible also for recommended policies.
+If you are not compatible, state what is lacking and what are your plans on how to achieve compliance.
+For current xSDK member packages: If you were not fully compatible at some point, please describe the steps you undertook to fulfill the policy. This information will be helpful for future xSDK member packages.
+
+**Website:** http://www.dealii.org
+
+### Mandatory Policies
+
+| Policy                 |Support| Notes                   |
+|------------------------|-------|-------------------------|
+|**M1.** Support xSDK community GNU Autoconf or CMake options. |Full| deal.II uses cmake.|
+|**M2.** Provide a comprehensive test suite for correctness of installation verification. |Full| deal.II has a testsuite that runs a large number of tests in debug mode, release mode, or both. As of early 2018, the total number of tests when all external dependencies are configured is around 9,500.|
+|**M3.** Employ userprovided MPI communicator (no MPI_COMM_WORLD). |Full| All objects take a communicator via the constructor. `MPI_COMM_WORLD` is only used as a default argument, but no part of the library assumes that things *have* to work on `MPI_COMM_WORLD`.  |
+|**M4.** Give best effort at portability to key architectures (standard Linux distributions, GNU, Clang, vendor compilers, and target machines at ALCF, NERSC, OLCF). |Full| We strive to build on all platforms, and test regularly with GNU, Clang, and other compilers. We consider the inability of building on a platform as a bug. |
+|**M5.** Provide a documented, reliable way to contact the development team. |Full| Mailing list addresses are provided at http://www.dealii.org . |
+|**M6.** Respect system resources and settings made by other previously called packages (e.g. signal handling). |Full| Yes. The only exception that comes to mind is that deal.II enables traps/exceptions when arithmetic operations encounter signaling NaNs. We use signaling NaNs in deal.II to default initialize objects with invalid values to ensure they cannot be used before receiving proper values.|
+|**M7.** Come with an open source (BSD style) license. |Full| LGPL. |
+|**M8.** Provide a runtime API to return the current version number of the software. |Full| Yes: `Utilities::dealii_version_string()`. |
+|**M9.** Use a limited and well-defined symbol, macro, library, and include file name space. |Full| All header files are prefixed with `deal.II/`. All C++ names are in a namespace `dealii::`. All macros are prefixed with `DEAL_II_`. |
```

Comment:
It might be worth mentioning that we overload common math functions (`std::pow`, `std::cos`, etc) for `VectorizedArray`. These are in `std::`.

Reply:
The statement 'All macros are prefixed with `DEAL_II_` is not true: `Assert`, `AssertThrow`, `DeclException`, etc. are all macros that do not satisfy this condition.

Reply:
Hm. Both comments are correct. I'll update the statement accordingly.

@drwells -- can you think of any other macros, other than the ones listed?

Reply:
I think I got all of them, but if we need everything I'll look more carefully:

```
./deal.II/base/exceptions.h:#define Assert(cond, exc)                                                   
./deal.II/base/exceptions.h:#define Assert(cond, exc)                                                   
./deal.II/base/exceptions.h:#define AssertNothrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define AssertNothrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define AssertThrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define AssertThrow(cond, exc)                                              
./deal.II/base/exceptions.h:#define DeclException0(Exception0)                                        
./deal.II/base/exceptions.h:#define DeclExceptionMsg(Exception, defaulttext)                          
./deal.II/base/exceptions.h:#define DeclException1(Exception1, type1, outsequence)                    
./deal.II/base/exceptions.h:#define DeclException2(Exception2, type1, type2, outsequence)             
./deal.II/base/exceptions.h:#define DeclException3(Exception3, type1, type2, type3, outsequence)      
./deal.II/base/exceptions.h:#define DeclException4(Exception4, type1, type2, type3, type4, outsequence) 
./deal.II/base/exceptions.h:#define DeclException5(Exception5, type1, type2, type3, type4, type5, outsequence) 
./deal.II/base/exceptions.h:#define DeclException0(Exception0)                                        
./deal.II/base/exceptions.h:#define DeclExceptionMsg(Exception, defaulttext)                          
./deal.II/base/exceptions.h:#define DeclException1(Exception1, type1, outsequence)                    
./deal.II/base/exceptions.h:#define DeclException2(Exception2, type1, type2, outsequence)             
./deal.II/base/exceptions.h:#define DeclException3(Exception3, type1, type2, type3, outsequence)      
./deal.II/base/exceptions.h:#define DeclException4(Exception4, type1, type2, type3, type4, outsequence) 
./deal.II/base/exceptions.h:#define DeclException5(Exception5, type1, type2, type3, type4, type5, outsequence) 
./deal.II/base/exceptions.h:#define AssertDimension(dim1,dim2) 
./deal.II/base/exceptions.h:#define AssertVectorVectorDimension(vec,dim1,dim2) 
./deal.II/base/exceptions.h:#define AssertIndexRange(index,range) 
./deal.II/base/exceptions.h:#define AssertIsFinite(number) 
./deal.II/base/exceptions.h:#define AssertThrowMPI(error_code) 
./deal.II/base/exceptions.h:#define AssertThrowMPI(error_code)
./deal.II/base/exceptions.h:#define AssertCuda(error_code) 
./deal.II/base/exceptions.h:#define AssertCusparse(error_code) 
./deal.II/sundials/arkode.h:#define AssertARKode(code) 
./deal.II/sundials/ida.h:#define AssertIDA(code) 
./deal.II/sundials/kinsol.h:#define AssertKINSOL(code) 
./deal.II/matrix_free/cuda_matrix_free.templates.h:#define BLOCK_SIZE 
./deal.II/matrix_free/cuda_matrix_free.templates.h:#define MAX_ELEM_DEGREE
```

Reply:
> It might be worth mentioning that we overload common math functions (std::pow, std::cos, etc) for VectorizedArray. These are in std::

Ditto for Adol-C ([`adolc_math.h`](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/adolc_math.h)) and some other customised functions for Sacado (['sacado_math.h'](https://github.com/dealii/dealii/blob/master/include/deal.II/differentiation/ad/sacado_math.h))

Reply:
I'm not worried about the `std::pow(...)` overloads -- they can't lead to ambiguities.

About `BLOCK_SIZE` and `MAX_ELEM_DEGREE`: @Rombur and @kronbichler, could these be renamed to something with the `DEAL_II_` prefix? Or could we actually make them into `constexpr` constants rather than preprocessor defines?

Reply:
We had an issue on the mailing list recently where a user wanted to do a matrix free projection with extra blocks. We should just put these types of things inside `numbers`.

Reply:
> About `BLOCK_SIZE` and `MAX_ELEM_DEGREE`

We should definitely change them. I think `static constexpr` variables should be fine for both. For `BLOCK_SIZE` it is immediately clear, for the latter it is used for an array to be declared a line below, so that should work as well. But I let someone with regular access to a GPU make the change - for me it would take a few days until I get a booking where I can compile with CUDA.

Reply:
I'll fix it. `BLOCK_SIZE` is an oversight, we were using this macros in several places and I thought I removed all of them but I guess I forgot one. I'll change `MAX_ELEM_DEGREE` too.

@drwells `BLOCK_SIZE` has nothing to do with `BlockVector` it is used to define how you lunch a kernel.